### PR TITLE
New method for getting element(')s respective attribute values

### PR DESCRIPTION
**Pull request by rkvinoth**
I would use it something like this:

```
from TM1py.Services import TM1Service

with TM1Service(address=ADDRESS, port=PORT, ssl=SSL, user=USER, password=PASSWORD) as tm1:
    period = tm1.elements.get_element_by_attribute(dimension_name='Period', hierarchy_name='Period', 
                                                   attribute='Month', elements=['01', '02'], 
                                                   exclude_empty_cells=True, element_unique_names=False)
```

the output would be:

`{'01': 'Jan', '02': 'Feb'}`

### CODE REVIEW

**Comment by rclapp** [comment]
You can do this already with the set mdx executor method

In the element service use execute_set_mdx

Pass it an MDX statement that resolves to the elements you want in this case {[period].[period].[02]}

Then specify what member or element properties you want. As a good test pass it ('*') to get back all properties, then you can fine tune it to only give you the content you want.

Ryan Clapp
AWS Business Systems
Sr. Manager

Sent from my mobile device

On Jul 9, 2020 12:48 AM, Vinoth Kumar Ravi <notifications@github.com> wrote:

I would use it something like this:

from TM1py.Services import TM1Service

with TM1Service(address=ADDRESS, port=PORT, ssl=SSL, user=USER, password=PASSWORD) as tm1:
    period = tm1.elements.get_element_by_attribute(dimension_name='Period', hierarchy_name='Period',
                                                   attribute='Month', elements=['01', '02'],
                                                   exclude_empty_cells=True, element_unique_names=False)


the output would be:

{'01': 'Jan', '02': 'Feb'}

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/cubewise-code/tm1py/pull/274

Commit Summary

  *   New method for getting element(')s respective attribute values

File Changes

  *   M TM1py/Services/ElementService.py<https://github.com/cubewise-code/tm1py/pull/274/files#diff-d805588ef7ff69bc98686bf09cd35030> (78)
  *   M Tests/Element.py<https://github.com/cubewise-code/tm1py/pull/274/files#diff-6d1568fea534d81de678a2ecb2a09eba> (23)

Patch Links:

  *   https://github.com/cubewise-code/tm1py/pull/274.patch
  *   https://github.com/cubewise-code/tm1py/pull/274.diff

-

**Comment by rkvinoth** [comment]
Just wanted a straight forward method. Will be definitely helpful for beginners. One good thing is that `elements` list is just a **Default** argument. No need of factoring or parsing anything.

For getting all elements:

```
from TM1py.Services import TM1Service

with TM1Service(address=ADDRESS, port=PORT, ssl=SSL, user=USER, password=PASSWORD) as tm1:
    period = tm1.elements.get_element_by_attribute(dimension_name='Period', hierarchy_name='Period', 
                                                   attribute='Month', exclude_empty_cells=True, 
                                                   element_unique_names=False)
```

File: TM1py/Services/ElementService.py [hunk 1]
```diff
@@ -168,11 +168,65 @@ def get_element_identifiers(self, dimension_name: str, hierarchy_name: str,
              """.format(
             elem_mdx=mdx_element_selection,
             attr_mdx=",".join(build_element_unique_names(
-                ["}ElementAttributes_" + dimension_name] * len(alias_attributes),
-                alias_attributes)),
+                ["}ElementAttributes_" + dimension_name] * len(alias_attributes), alias_attributes)),
             dim=dimension_name)
         return self._retrieve_mdx_rows_and_cell_values_as_string_set(mdx, **kwargs)
 
+    def get_element_by_attribute(self, dimension_name: str, hierarchy_name: str, attribute: str,
```
Comment (MariusWirtz) [inline TM1py/Services/ElementService.py:175]: I think the name sounds too much like filtering.

I think a name like `get_attribute_of_elements` would be more indicative.

File: TM1py/Services/ElementService.py [hunk 2]
```diff
@@ -168,11 +168,65 @@ def get_element_identifiers(self, dimension_name: str, hierarchy_name: str,
...
+                                 elements: Union[str, List[str]] = None, **kwargs) -> dict:
+        """
+         Get element name and attribute value for a set of elements in a hierarchy
+
+        :param dimension_name:
+        :param hierarchy_name:
+        :param attribute: Name of the Attribute
+        :param elements:  MDX (Set) expression or iterable of elements
+        :keyword exclude_empty_cells: Boolean
+        :keyword element_unique_names: Boolean
```
Comment (MariusWirtz) [inline TM1py/Services/ElementService.py:185]: Why use `kwargs` for these two arguments?

I would prefer to have them as optional arguments with default values. Then PyCharm, and other smart IDEs, would recognize them as arguments to the function.
Currently, we use `kwargs` almost exclusively for "technical" arguments that are consistent throughout the entire library, such as, `timeout`

File: TM1py/Services/ElementService.py (hunk 1 above)
Comment (rkvinoth) [reply TM1py/Services/ElementService.py]: I like this name.

File: TM1py/Services/ElementService.py (hunk 2 above)
Comment (rkvinoth) [reply TM1py/Services/ElementService.py]: yeah makes more sense to make them Default.

**Comment by MariusWirtz** [comment]
Thanks for your contribution!

**Pull request merged**